File: cli/internal/retrieval/retrieval.go
package retrieval

import (
	"sort"
	"strings"
	"unicode"
)

type Item struct {
	Term       string `json:"term"`
	TargetPath string `json:"targetPath"`
}

type Options struct {
	TopN int
}

type Match struct {
	Term                string   `json:"term"`
	TargetPath          string   `json:"targetPath"`
	NormalizedTerm      string   `json:"normalizedTerm"`
	Score               float64  `json:"score"`
	Exact               bool     `json:"exact"`
	MatchType           string   `json:"matchType"`
	TermRuneLen         int      `json:"termRuneLen"`
	BigramCoverage      float64  `json:"bigramCoverage"`
	TrigramCoverage     float64  `json:"trigramCoverage"`
	MatchedBigramCount  int      `json:"matchedBigramCount"`
	MatchedTrigramCount int      `json:"matchedTrigramCount"`
	MatchedBigrams      []string `json:"matchedBigrams"`
	MatchedTrigrams     []string `json:"matchedTrigrams"`
}

func NormalizeChinese(s string) string {
	var b strings.Builder
	for _, r := range s {
		switch {
		case r == '\u3000':
			continue
		case r >= '\uff01' && r <= '\uff5e':
			r -= '\ufee0'
		}
		r = unicode.ToLower(r)
		if unicode.Is(unicode.Han, r) || unicode.IsLetter(r) || unicode.IsDigit(r) {
			b.WriteRune(r)
		}
	}
	return b.String()
}

func Ngrams(s string, n int) []string {
	if n <= 0 {
		return []string{}
	}
	runes := []rune(s)
	if len(runes) < n {
		return []string{}
	}
	grams := make([]string, 0, len(runes)-n+1)
	for i := 0; i <= len(runes)-n; i++ {
		grams = append(grams, string(runes[i:i+n]))
	}
	return grams
}

func Search(items []Item, question string, opts Options) []Match {
	normalizedQuestion := NormalizeChinese(question)
	if normalizedQuestion == "" {
		return []Match{}
	}
	queryBigrams := ngramSet(Ngrams(normalizedQuestion, 2))
	queryTrigrams := ngramSet(Ngrams(normalizedQuestion, 3))
	byTarget := map[string]Match{}
	for _, item := range items {
		match, ok := scoreItem(item, normalizedQuestion, queryBigrams, queryTrigrams)
		if !ok {
			continue
		}
		if existing, ok := byTarget[match.TargetPath]; !ok || matchLess(match, existing) {
			byTarget[match.TargetPath] = match
		}
	}
	matches := make([]Match, 0, len(byTarget))
	for _, match := range byTarget {
		matches = append(matches, match)
	}
	sort.SliceStable(matches, func(i, j int) bool {
		return matchLess(matches[i], matches[j])
	})
	matches = suppressContainedTerms(matches)
	if opts.TopN > 0 && len(matches) > opts.TopN {
		matches = matches[:opts.TopN]
	}
	return matches
}

func scoreItem(item Item, normalizedQuestion string, queryBigrams, queryTrigrams map[string]bool) (Match, bool) {
	normalizedTerm := NormalizeChinese(item.Term)
	if normalizedTerm == "" || item.TargetPath == "" {
		return Match{}, false
	}
	termRuneLen := len([]rune(normalizedTerm))
	termBigrams := Ngrams(normalizedTerm, 2)
	termTrigrams := Ngrams(normalizedTerm, 3)
	matchedBigrams, bigramCoverage := coverage(termBigrams, queryBigrams)
	matchedTrigrams, trigramCoverage := coverage(termTrigrams, queryTrigrams)
	match := Match{
		Term:                item.Term,
		TargetPath:          item.TargetPath,
		NormalizedTerm:      normalizedTerm,
		TermRuneLen:         termRuneLen,
		BigramCoverage:      bigramCoverage,
		TrigramCoverage:     trigramCoverage,
		MatchedBigramCount:  len(matchedBigrams),
		MatchedTrigramCount: len(matchedTrigrams),
		MatchedBigrams:      matchedBigrams,
		MatchedTrigrams:     matchedTrigrams,
	}
	if strings.Contains(normalizedQuestion, normalizedTerm) {
		match.Exact = true
		match.MatchType = "exact"
		match.Score = 10000 + float64(termRuneLen*100)
		return match, true
	}
	if termRuneLen <= 2 {
		return Match{}, false
	}
	switch {
	case termRuneLen == 3:
		if bigramCoverage < 1.0 {
			return Match{}, false
		}
	default:
		if len(matchedBigrams) < 2 || bigramCoverage < 0.5 {
			return Match{}, false
		}
	}
	match.MatchType = "fuzzy"
	match.Score = bigramCoverage*100 + trigramCoverage*80 + float64(len(matchedBigrams))*5 + float64(len(matchedTrigrams))*8 + float64(termRuneLen)
	return match, true
}

func coverage(grams []string, query map[string]bool) ([]string, float64) {
	unique := uniqueOrdered(grams)
	if len(unique) == 0 {
		return []string{}, 0
	}
	matched := make([]string, 0, len(unique))
	for _, gram := range unique {
		if query[gram] {
			matched = append(matched, gram)
		}
	}
	return matched, float64(len(matched)) / float64(len(unique))
}

func ngramSet(grams []string) map[string]bool {
	set := make(map[string]bool, len(grams))
	for _, gram := range grams {
		set[gram] = true
	}
	return set
}

func uniqueOrdered(values []string) []string {
	seen := map[string]bool{}
	out := make([]string, 0, len(values))
	for _, value := range values {
		if seen[value] {
			continue
		}
		seen[value] = true
		out = append(out, value)
	}
	return out
}

func suppressContainedTerms(matches []Match) []Match {
	filtered := make([]Match, 0, len(matches))
	for _, match := range matches {
		containedBySelected := false
		for _, selected := range filtered {
			switch {
			case match.NormalizedTerm == selected.NormalizedTerm:
			case strings.Contains(selected.NormalizedTerm, match.NormalizedTerm):
				containedBySelected = true
			case selected.Exact && !match.Exact && strings.Contains(match.NormalizedTerm, selected.NormalizedTerm):
				containedBySelected = true
			case selected.Exact && !match.Exact && len(match.MatchedBigrams) > 0 && containsAll(selected.MatchedBigrams, match.MatchedBigrams):
				containedBySelected = true
			}
			if containedBySelected {
				break
			}
		}
		if containedBySelected {
			continue
		}
		filtered = append(filtered, match)
	}
	return filtered
}

func containsAll(values, subset []string) bool {
	valueSet := map[string]bool{}
	for _, value := range values {
		valueSet[value] = true
	}
	for _, value := range subset {
		if !valueSet[value] {
			return false
		}
	}
	return true
}

func matchLess(a, b Match) bool {
	if a.Exact != b.Exact {
		return a.Exact
	}
	if a.Score != b.Score {
		return a.Score > b.Score
	}
	if a.TermRuneLen != b.TermRuneLen {
		return a.TermRuneLen > b.TermRuneLen
	}
	if a.Term != b.Term {
		return a.Term < b.Term
	}
	return a.TargetPath < b.TargetPath
}

File: packages/data-harness-cli/src/lib/retrieval.js
export function normalizeChinese(s) {
  let out = "";
  for (const r of s) {
    let code = r.codePointAt(0);
    if (code === 0x3000) continue;
    if (code >= 0xff01 && code <= 0xff5e) code -= 0xfee0;
    const ch = String.fromCodePoint(code).toLowerCase();
    if (/\p{Script=Han}|\p{L}|\p{N}/u.test(ch)) out += ch;
  }
  return out;
}

export function ngrams(s, n) {
  if (n <= 0) return [];
  const runes = [...s];
  if (runes.length < n) return [];
  const grams = [];
  for (let i = 0; i <= runes.length - n; i++) grams.push(runes.slice(i, i + n).join(""));
  return grams;
}

export function search(items, question, opts = {}) {
  const normalizedQuestion = normalizeChinese(question);
  if (!normalizedQuestion) return [];
  const queryBigrams = ngramSet(ngrams(normalizedQuestion, 2));
  const queryTrigrams = ngramSet(ngrams(normalizedQuestion, 3));
  const byTarget = new Map();
  for (const item of items) {
    const match = scoreItem(item, normalizedQuestion, queryBigrams, queryTrigrams);
    if (!match) continue;
    const existing = byTarget.get(match.targetPath);
    if (!existing || matchLess(match, existing)) byTarget.set(match.targetPath, match);
  }
  let matches = [...byTarget.values()];
  matches.sort((a, b) => (matchLess(a, b) ? -1 : matchLess(b, a) ? 1 : 0));
  matches = suppressContainedTerms(matches);
  const topN = opts.topN || 0;
  if (topN > 0 && matches.length > topN) matches = matches.slice(0, topN);
  return matches;
}

function scoreItem(item, normalizedQuestion, queryBigrams, queryTrigrams) {
  const normalizedTerm = normalizeChinese(item.term);
  if (!normalizedTerm || !item.targetPath) return null;
  const termRuneLen = [...normalizedTerm].length;
  const termBigrams = ngrams(normalizedTerm, 2);
  const termTrigrams = ngrams(normalizedTerm, 3);
  const [matchedBigrams, bigramCoverage] = coverage(termBigrams, queryBigrams);
  const [matchedTrigrams, trigramCoverage] = coverage(termTrigrams, queryTrigrams);
  const match = {
    term: item.term,
    targetPath: item.targetPath,
    normalizedTerm,
    termRuneLen,
    bigramCoverage,
    trigramCoverage,
    matchedBigramCount: matchedBigrams.length,
    matchedTrigramCount: matchedTrigrams.length,
    matchedBigrams,
    matchedTrigrams,
    exact: false,
    matchType: "",
    score: 0,
  };
  if (normalizedQuestion.includes(normalizedTerm)) {
    match.exact = true;
    match.matchType = "exact";
    match.score = 10000 + termRuneLen * 100;
    return match;
  }
  if (termRuneLen <= 2) return null;
  if (termRuneLen === 3) {
    if (bigramCoverage < 1.0) return null;
  } else if (matchedBigrams.length < 2 || bigramCoverage < 0.5) {
    return null;
  }
  match.matchType = "fuzzy";
  match.score =
    bigramCoverage * 100 +
    trigramCoverage * 80 +
    matchedBigrams.length * 5 +
    matchedTrigrams.length * 8 +
    termRuneLen;
  return match;
}

function coverage(grams, query) {
  const unique = uniqueOrdered(grams);
  if (unique.length === 0) return [[], 0];
  const matched = unique.filter((gram) => query.has(gram));
  return [matched, matched.length / unique.length];
}

function ngramSet(grams) {
  return new Set(grams);
}

function uniqueOrdered(values) {
  const seen = new Set();
  const out = [];
  for (const value of values) {
    if (seen.has(value)) continue;
    seen.add(value);
    out.push(value);
  }
  return out;
}

function suppressContainedTerms(matches) {
  const filtered = [];
  for (const match of matches) {
    let containedBySelected = false;
    for (const selected of filtered) {
      if (match.normalizedTerm === selected.normalizedTerm) continue;
      if (selected.normalizedTerm.includes(match.normalizedTerm)) containedBySelected = true;
      else if (selected.exact && !match.exact && match.normalizedTerm.includes(selected.normalizedTerm)) {
        containedBySelected = true;
      } else if (
        selected.exact &&
        !match.exact &&
        match.matchedBigrams.length > 0 &&
        containsAll(selected.matchedBigrams, match.matchedBigrams)
      ) {
        containedBySelected = true;
      }
      if (containedBySelected) break;
    }
    if (!containedBySelected) filtered.push(match);
  }
  return filtered;
}

function containsAll(values, subset) {
  const valueSet = new Set(values);
  return subset.every((value) => valueSet.has(value));
}

function matchLess(a, b) {
  if (a.exact !== b.exact) return a.exact;
  if (a.score !== b.score) return a.score > b.score;
  if (a.termRuneLen !== b.termRuneLen) return a.termRuneLen > b.termRuneLen;
  if (a.term !== b.term) return a.term < b.term;
  return a.targetPath < b.targetPath;
}
